cpp/src/arrow/compute/kernels/take-internal.h

```diff
@@ -455,20 +493,181 @@ class TakerImpl<IndexSequence, StructType> : public Taker<IndexSequence> {
+    RETURN_NOT_OK(null_bitmap_builder_->Reserve(indices.length()));
+    RETURN_NOT_OK(type_id_builder_->Reserve(indices.length()));
+    RETURN_NOT_OK(VisitIndices(indices, values, [&](int64_t index, bool is_valid) {
+      null_bitmap_builder_->UnsafeAppend(is_valid);
+      type_id_builder_->UnsafeAppend(type_ids[index]);
+      child_counts[type_ids[index]] += is_valid;
+      return Status::OK();
+    }));
+
+    if (union_type_->mode() == UnionMode::SPARSE) {
+      // bounds checking was done while appending to the null bitmap
+      indices.set_never_out_of_bounds();
+
+      for (int i = 0; i < this->type_->num_children(); ++i) {
+        RETURN_NOT_OK(sparse_children_[i]->Take(*union_array.child(i), indices));
+      }
+    } else {
+      // Gathering from the offsets into child arrays is a bit tricky.
+      // Allocate temporary storage for the offsets of all valid slots
+      std::shared_ptr<Buffer> child_offsets_storage;
+      RETURN_NOT_OK(AllocateBuffer(
+          pool_, (indices.length() - indices.null_count()) * sizeof(int32_t),
+          &child_offsets_storage));
+
+      // Partition offsets by type_id: child_offset_partitions[type_id] will
+      // point to storage for child_counts[type_id] offsets
+      std::vector<int32_t*> child_offset_partitions(child_counts.size());
+      auto child_offsets_storage_data = GetInt32(child_offsets_storage);
+      for (auto type_id : union_type_->type_codes()) {
+        child_offset_partitions[type_id] = child_offsets_storage_data;
+        child_offsets_storage_data += child_counts[type_id];
+      }
+
+      // Fill child_offsets_storage with the taken offsets
+      RETURN_NOT_OK(offset_builder_->Reserve(indices.length()));
+      RETURN_NOT_OK(VisitIndices(indices, values, [&](int64_t index, bool is_valid) {
+        auto type_id = type_ids[index];
+        if (is_valid) {
+          offset_builder_->UnsafeAppend(child_length_[type_id]++);
+          *child_offset_partitions[type_id] = union_array.value_offset(index);
+          ++child_offset_partitions[type_id];
+        } else {
+          offset_builder_->UnsafeAppend(0);
+        }
+        return Status::OK();
+      }));
+
+      // Take from each child at those offsets
+      int64_t taken_offset_begin = 0;
+      for (int i = 0; i < this->type_->num_children(); ++i) {
+        auto type_id = union_type_->type_codes()[i];
+        auto length = child_counts[type_id];
+        Int32Array taken_offsets(length, SliceBuffer(child_offsets_storage,
+                                                     sizeof(int32_t) * taken_offset_begin,
+                                                     sizeof(int32_t) * length));
+        ArrayIndexSequence<Int32Type> child_indices(taken_offsets);
+        child_indices.never_out_of_bounds();
+        RETURN_NOT_OK(dense_children_[i]->Take(*union_array.child(i), child_indices));
+        taken_offset_begin += length;
+      }
+    }
```

Comment:
Add something like `DCHECK_EQ(taken_offset_begin * sizeof(int32_t), child_offsets_storage->size())`?

Reply:
`child_offsets_storage->size()` might be larger than that if the union array also has nulls

Reply:
Shouldn't the sum of all `child_counts` be equal to `(indices.length() - indices.null_count()) * sizeof(int32_t)`, which is also the size of the `child_offsets_storage`?

Reply:
Exactly